### PR TITLE
Test: trigger AfterFailed before AfterEach when is in Context

### LINKED ISSUE #3481: CI: some tests remove resources before logs are gathered
Within Ginkgo, there is the notion of `Context`, which allows subsets of tests to occur within a `Describe`. For instance, Within a `Describe`, one can have the following:

- Describe
    - BeforeEach (for each `It` within Describe)
    - AfterEach (for each `It` within Describe)
    - JustAfterEach
    - AfterFailed
    - It
    - Context
        - BeforeEach (for each `It` within Context)
        - AfterEach (for each `It` within Context)
            - It 

For the 'It' within the context, the following would occur if a test fails:

BeforeEach (for each `It` within Describe) -->  BeforeEach (for each `It` within Context) --> It --> AfterEach (for each `It` within Context) --> JustAfterEach --> AfterFailed --> AfterEach (for each `It` within Describe) 
  

This means that if within the `AfterEach` for the `Context`, if we delete any resources provisioned within the `Context`, but gather logs in the `Describe`'s `AfterEach`, we will have no resources to analyze. This can be solved by moving the `JustAfterEach` and `AfterFailed` to be within each `Context`, and if a `Describe` has a `Context` within it, ensuring that all `It`s are covered by a `Context` which contains `AfterEach` and `AfterFailed` within it. 

Tests that need fixing:
- [ ] `test/runtime/connectivity.go:RuntimeValidatedConnectivityTest`

TODO: need to go through all tests and see what are affected by this.

### PR DESCRIPTION
**Pull request by eloycoto**
When a context is defined with a AfterEach the AfterEach function will
be called before `JustAfterEach` and `AfterFailed` and some info cannot
be retrieved correctly because no longer exits. With this commit all the
`JustAfterEach` and `AfterFailed` will be called just before the
AfterEach.

Fix #3481

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by ianvernon** [comment]
I am not a fan of this. It subverts how Ginkgo normally functions. We can just have `AfterFailed` and `JustAfterEach` within each context and it solves the problem without having to run things outside of the scope they were declared. Why can't we do that?

**Comment by eloycoto** [comment]
Hi @ianvernon 

I think that it's the only way that we can ensure that all works as we expected.
The idea of having all as ginkgo, it's good but Ginkgo does not want to have
something that we already need and that was requested in the past months, like
`BeforeAll`,`AfterAll`, `AfterFailed`, etc..  Ginkgo issue #70. We already
changed the ginkgo behaviour to Cilium behaviour, as other projects already
did.


The main issue is, as you wrote in the issue, your code will be like this one:

```
Describe
    BeforeEach (for each `It` within Describe)
    AfterEach (for each `It` within Describe)
    JustAfterEach
    AfterFailed
    It
    Context
        BeforeEach (for each `It` within Context)
        AfterEach (for each `It` within Context)
        It
```

In this case, the `JustAfterEach` will execute once per each It, as you propose
your solution will be something like this:

```
Describe
    BeforeEach (for each `It` within Describe)
    AfterEach (for each `It` within Describe)
    JustAfterEach
    AfterFailed
    It
    Context
        BeforeEach (for each `It` within Context)
        AfterEach (for each `It` within Context)
        JustAfterEach
        AfterFailed
        It
```

The main problem with this code, it's that the `AfterFailed` will happens two
times instead of one and the second one will *delete* all the information gather
by the first one, so that solution is not valid at all.

The other option is to write the code in a different way, like:
```

Describe
    Context A
        BeforeEach (for each `It` within Describe)
        AfterEach (for each `It` within Describe)
        JustAfterEach
        AfterFailed
        It
    Context
        BeforeEach (for each `It` within Context)
        AfterEach (for each `It` within Context)
        JustAfterEach
        AfterFailed
        It
```

In this case the `AfterFailed` will happens only one time, but all the test need
to be rewrited to have all of this structure. With this situation a lot of
boilerplate code will be duplicated, at the moment we call AfterFailed 25 times,
we should created another 19 AfterFailed/JustAfterEach, that is where the
Context is now defined in the text. Add the same code 19 times, and be able to
replace some data because a global AfterFailed is in place is not my preference
TBH, thing will get more verbose, and difficult to debug.

Other solution that you may have is to use `Sync.Once` on `AfterFailed`
function, but that it's buggy too. If one test has failed in the given Describe
or Context the next test that fails will not gather data. This can work in the
PR with the `failfast enable`, but it will not work on Nigthly test and in
master branch test.

Change Ginkgo behaviour is not my ideal too, but they had strong ideals about
what test framework should do, this does not fit 100% for us. At the moment, we
already changed Ginkgo code. This 46 lines taking care of the behaviour that
reduces some boilerplate code, and helps us to avoid duplicated code and add two
functionalities that are really useful for us.

Waiting for your reply.
Eloy

File: test/ginkgo-ext/scopes.go [hunk 1]
```diff
@@ -151,23 +151,65 @@ func AfterFailed(body func()) bool {
 	return true
 }
 
-// RunAfterEach is a wrapper that executes all AfterEach functions that are
-// stored in cs.afterEach array.
-func RunAfterEach(cs *scope) {
-	if cs == nil {
+// justAfterEachStatus map to store what `justAfterEach` functions have been
+// already executed for the given test
+var justAfterEachStatus map[string]bool = map[string]bool{}
```
Comment (aanm) [inline test/ginkgo-ext/scopes.go:156]: Do we run parallel tests? If yes please use a mutex for this map

Comment (eloycoto) [reply test/ginkgo-ext/scopes.go:156]: No, we don't run parallel test. The parallel test happens in Jenkins.

**Comment by aanm** [comment]
@ianvernon can I merge this?

**Comment by ianvernon** [comment]
Sorry for the delay in getting back to you on this. I still disagree with this implementation. I'll respond inline to your statements.

> The idea of having all as ginkgo, it's good but Ginkgo does not want to have
something that we already need and that was requested in the past months, like
BeforeAll,AfterAll, AfterFailed, etc.. Ginkgo issue #70. We already
changed the ginkgo behaviour to Cilium behaviour, as other projects already
did.

This is true, but to me, the entire point of `Context` is to scope certain operations. Your change breaks that scoping at a fundamental level. The addition of things like `BeforeAll` or `AfterAll` augments the original framework instead of changing the fundamental behavior of the framework. The original constructs (`BeforeEach`, `AfterEach`, `Context`), still work the same way they did before with `BeforeAll` and `AfterAll`. `AfterFailed` is basically a wrapper which catches failures and performs certain operations. I still view this as augmenting, not altering, the behavior of the framework. But, here you are changing the fundamental behavior of `Context`. 

> The main problem with this code, it's that the AfterFailed will happens two
times instead of one and the second one will delete all the information gather
by the first one, so that solution is not valid at all.

We don't have to delete all of the information? I'm fairly sure we can easily fix this in our helper functions.

> In this case the AfterFailed will happens only one time, but all the test need
to be rewrited to have all of this structure. With this situation a lot of
boilerplate code will be duplicated, at the moment we call AfterFailed 25 times,
we should created another 19 AfterFailed/JustAfterEach, that is where the
Context is now defined in the text. Add the same code 19 times, and be able to
replace some data because a global AfterFailed is in place is not my preference
TBH, thing will get more verbose, and difficult to debug.

I personally don't see a problem with this. It makes it perfectly clear what behavior is scoped to each context. Verbosity isn't necessarily a bad thing. 

> Other solution that you may have is to use Sync.Once on AfterFailed
function, but that it's buggy too. If one test has failed in the given Describe
or Context the next test that fails will not gather data. This can work in the
PR with the failfast enable, but it will not work on Nigthly test and in
master branch test.

I don't understand what you mean here. If we have the `AfterFailed` within each context, we can customize the behavior per context to do different things depending upon the context; that's the whole point of the context in the first place. If the `It` within the `Context` fails, don't we create a directory with logs with `DescribeNameContextNameItName`?Why exactly would the next test that fails not gather data? Please explain further or correct me if I am misinterpreting you here.

Overall I still don't like this change. I do not think that your argument that the fact that we augmented the existing framework means we can change fundamental behavior of some of its constructs is valid. I also do not always think that verbosity is necessarily a bad thing, as verbosity can be useful in that it clearly delineates what operation is happening in a specific scope.

**Comment by eloycoto** [comment]
Hi,

I'm not sure that you understand the issue, you still see `JustAfterEach` and
`AfterFailed` as a corellation of functions that spilt the `afterEach`, but
these methors have the same behaviour as the `JustBeforeEach`(Oficial method
from Ginkgo) as example:

```
Describe
    BeforeEach(A)
    JustBeforeEach(A)
    Context
        BeforeEach(B)
        IT(TEST1)

```

The execution will be like:

```
Describe
BeforeEach(A)
BeforeEach(B)
JustBeforeEach(A)
IT(TEST1)
```

Another example:

```
Describe
    BeforeEach(A)
    JustBeforeEach(A)
    Context
        BeforeEach(B)
        JustBeforeEach(B)
        IT(TEST1)

```
Execution:


```
Describe
BeforeEach(A)
BeforeEach(B)
JustBeforeEach(A)
JustBeforeEach(B)
IT(TEST1)
```

It's a `beforeEach` that will happens **after** the `BeforeEach` but it's not in
the same scope. If you see, is the same behaviour that we're using in the
`JustAfterEach`.

As your recommend, the 19 times that we call with the `AfterFailed`, are always
the same, with your proposal you'll add 15 more, that will make exactly the
same thing, and *Do not follow* how ginkgo works, as you had seen, the
`JustAfterEach()` is the same before as `JustBeforeEach()`

In the other hand, it'll be buggy, let's get an example from
`runtime/Policies.go`, where we have test in a Context and other that are not
in the `context`, the flow will be like this:

Test Group description:
```
Describe
BeforeAll(A)
AfterAll(A)
AfterFailed(A)
AfterEach(A)
JustAfterEach(A)
TEST1
TEST2
TEST3
Context
    BeforeAll(B)
    AfterAll(B)
    AfterFailed(B)
    AfterEach(B)
    JustAfterEach(B)
    TESTB1
    TESTB2
    TESTB3
```

In this case, the execution will be like this:

```
Describe
BeforeAll
TEST1; JustAfterEach(A), AfterFailed(A), AfterEach(A)
TEST2; JustAfterEach(A), AfterFailed(A), AfterEach(A)
TEST3; JustAfterEach(A), AfterFailed(A), AfterEach(A)
Context
    BeforeAll(A)
    TESTB1:
       JustAfterEach(B)
       AfterFailed(B)
       AfterEach(B)
       JustAfterEach(A)
       AfterFailed(A) ❌ Here the vm.ReportFailed will delete the exiting data created by AfterFailed(A) and create a new one, without the information that AfterEach Do.
       AfterEAch(A)
    TESTB2:
       JustAfterEach(B)
       AfterFailed(B)
       AfterEach(B)
       JustAfterEach(A)
       AfterFailed(A) ❌ Here the vm.ReportFailed will delete the exiting data created by AfterFailed(A) and create a new one, without the information that AfterEach Do.
       AfterEAch(A)
    TESTB3:
       JustAfterEach(B)
       AfterFailed(B)
       AfterEach(B)
       JustAfterEach(A)
       AfterFailed(A) ❌ Here the vm.ReportFailed will delete the exiting data created by AfterFailed(A) and create a new one, without the information that AfterEach Do.
       AfterEAch(A)
    AfterAll(B)
AfterAll(A)
```

For me, this approach with the AfterFailed, JustAfter doesn't make sense at all. With my changes will be like this:


```
Describe
BeforeAll
TEST1; JustAfterEach(A), AfterFailed(A), AfterEach(A)
TEST2; JustAfterEach(A), AfterFailed(A), AfterEach(A)
TEST3; JustAfterEach(A), AfterFailed(A), AfterEach(A)
Context
    BeforeAll(A)
    TESTB1:
       JustAfterEach(B); JustAfterEach(A)
       AfterFailed(B); AfterFailed(A);
       AfterEach(B) ; AfterEach(A);
    TESTB2:
       JustAfterEach(B); JustAfterEach(A)
       AfterFailed(B); AfterFailed(A);
       AfterEach(B) ; AfterEach(A);
    TESTB3:
       JustAfterEach(B); JustAfterEach(A)
       AfterFailed(B); AfterFailed(A);
       AfterEach(B) ; AfterEach(A);
    AfterAll(B)
AfterAll(A)
```

Plus, if the `AfterFailed` is the same, *as it is now*, the times that the
`AfterFailed` is defined it's only once per given Describe. Futhermore for
someone that don't use to write test, only need to know that the
`JustAfterEach`  and `AfterFailed` happens before all `AfterEach` instead of
need to know a special behaviour that **it's not similar to what Ginkgo uses to do**.

I'm happy to refactor all the test, but I think that we're adding a new layer
of complexity that it's not needed at all. IMHO

Kind regards.
Eloy

**Comment by ianvernon** [comment]
@eloycoto OK, now I see what you mean! Thank you for providing some more examples. You are correct, I was not fully understanding. Can you please add documentation that illustrates this functionality in `Documentation/contributing.rst`? It will be helpful for those who are knew to the framework to understand how the separate constructs work in tandem with one another.

**Comment by ianvernon** [comment]
@eloycoto do you think this should be backported? It will help us get logs for bugs that might be in the 1.0 branch.

**Review by ianvernon — APPROVED** [review]

**Review by ianvernon — APPROVED** [review]

**Comment by eloycoto** [comment]
yes, I think that should be backported.

**Pull request merged**